File: LeetCode/0200-number-of-islands/0200-number-of-islands.swift
class Solution {
    func numIslands(_ grid: [[Character]]) -> Int {
        let rowCount = grid.count
        let colCount = grid[0].count
        var visited = Array(repeating: Array(repeating: false, count: colCount), count: rowCount)
        var result = 0
        
        for row in 0..<rowCount {
            for col in 0..<colCount {
                if grid[row][col] == "1" && visited[row][col] == false {
                    bfs(row, col)
                    result += 1
                }
            }
        }
        
        func bfs(_ row: Int, _ col: Int) {
            let dRow = [-1, 1, 0, 0]
            let dCol = [0, 0, -1, 1]
            var queue = [(row, col)]
            
            while !queue.isEmpty {
                let (curRow, curCol) = queue.removeFirst()
                
                for i in 0..<4 {
                    let nextRow = curRow + dRow[i]
                    let nextCol = curCol + dCol[i]
                    
                    if nextRow >= 0 && nextRow < rowCount && nextCol >= 0 && nextCol < colCount {
                        if grid[nextRow][nextCol] == "1" && visited[nextRow][nextCol] == false {
                            queue.append((nextRow, nextCol))
                            visited[nextRow][nextCol] = true

                        }
                    }
                }
            }
        }
        
        return result
    }
}

File: LeetCode/0200-number-of-islands/0200-number-of-islands.py
from collections import deque

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        result = 0
        rowCount = len(grid)
        colCount = len(grid[0])
        visited = [[False] * colCount for _ in range(rowCount)]

        def bfs(row, col):
            dRow = [-1, 1, 0 ,0]
            dCol = [0, 0, -1, 1]
            visited[row][col] = True
            queue = deque()
            queue.append((row, col))

            while queue:
                curRow, curCol = queue.popleft()
                for i in range(4):
                    nextRow = curRow + dRow[i]
                    nextCol = curCol + dCol[i]

                    if nextRow >= 0 and nextRow < rowCount and nextCol >= 0 and nextCol < colCount:
                        if grid[nextRow][nextCol] == '1' and not visited[nextRow][nextCol]:
                            visited[nextRow][nextCol] = True
                            queue.append((nextRow, nextCol))
        
        for i in range(rowCount):
            for j in range(colCount):
                if grid[i][j] == '1' and visited[i][j] == False:
                    bfs(i, j)
                    result += 1
        return result